operator/pkg/resources/statefulset.go

```diff
@@ -1018,66 +1014,6 @@ func (r *StatefulSetResource) AdditionalListenersEnvVars() []corev1.EnvVar {
 	}}
 }
 
-// GetPortsForListenersInAdditionalConfig gets the ports for the additional listeners and advertised APIs set in addtionalConfiguration.
-// - redpanda.kafka_api
-// - redpanda.advertised_kafka_api
-// - pandaproxy.pandaproxy_api
-// - pandaproxy.advertised_pandaproxy_api
-// example: redpanda.kafka_api: "[{'name':'private-link','address':'0.0.0.0','port':39002}]"
-// example: redpanda.advertised_kafka_api: "[{'name':'private-link','address':'{{ .Index }}-f415bda0-{{ .HostIP | sha256sum | substr 0 7 }}.cluster123.fmc.prd.cloud.redpanda.com','port': 'port': {{30092 | add .Index}}}]"
-func (r *StatefulSetResource) GetPortsForListenersInAdditionalConfig() []corev1.ContainerPort {
-	ports := []corev1.ContainerPort{}
-
-	if len(r.pandaCluster.Spec.AdditionalConfiguration) == 0 {
-		return ports
-	}
-
-	additionalNode0Config := config.ProdDefault()
-	for _, k := range additionalListenerCfgNames {
-		if v, found := r.pandaCluster.Spec.AdditionalConfiguration[k]; found {
-			res, err := utils.Compute(v, utils.NewEndpointTemplateData(0, "dummy", 0), false)
-			if err != nil {
-				r.logger.Error(err, "failed to evaluate template", "template", v)
-				continue
-			}
-			err = config.Set(additionalNode0Config, k, res)
-			if err != nil {
-				r.logger.Error(err, "failed to set node config", k, v)
-				continue
-			}
-		}
-	}
-
-	for i := 0; i < int(ptr.Deref(r.nodePool.Replicas, 0)); i++ {
-		for _, n := range additionalNode0Config.Redpanda.AdvertisedKafkaAPI {
-			ports = append(ports, corev1.ContainerPort{
-				Name:          getAdditionalListenerPortName(n.Name, i),
-				ContainerPort: int32(n.Port + i),
-			})
-		}
-		if additionalNode0Config.Pandaproxy != nil {
-			for _, n := range additionalNode0Config.Pandaproxy.AdvertisedPandaproxyAPI {
-				ports = append(ports, corev1.ContainerPort{
-					Name:          getAdditionalListenerPortName(n.Name, i),
-					ContainerPort: int32(n.Port + i),
-				})
-			}
-		}
-	}
```

Comment:
I think the function is still needed, and the code should be changed to something like,
```
                for i, n := range additionalNode0Config.Redpanda.KafkaAPI {
                        ports = append(ports, corev1.ContainerPort{
                                Name:          getAdditionalListenerPortName(n.Name, i),
                                ContainerPort: int32(n.Port),
                        })
                }
                if additionalNode0Config.Pandaproxy != nil {
                        for i, n := range additionalNode0Config.Pandaproxy.PandaproxyAPI {
                                ports = append(ports, corev1.ContainerPort{
                                        Name:          getAdditionalListenerPortName(n.Name, i),
                                        ContainerPort: int32(n.Port),
                                })
                        }
                }
```
It depends on PL tests.
Ultimately we need to open the port that RP broker listens to for Kafka API and proxy.  Additional Kafka listener can have different port than internal (9092) and default external listener (30092), we need to open it up. I hope that K8s statefulset controller will consolidate the container ports if duplications.

Reply:
@paulzhang97  updated.
i also had to change some semantics, i found out that we actually never add the correct port, only the advertised port, as these seem to override each other. so in the end, the advertised port was added to container ports, but it's useless, because nothing listens at this port (it's only a kafka api thing, and for the load balancer).

the definitions of advertised and port in additionalProperties cancelled out each other. but it did not matter, as ports work, even if not declared in container ports.

in the end, it would be much better to properly define listeners in spec.configuration.kafkaAPI, instead of "hacking in" these listeners via env vars passed into the configurator. but this goes beyond this PR.

Reply:
@paulzhang97 AFAIK it's because `containerPort` can override the container's exposed ports at runtime. The `redpanda` container image has the following ports exposed by default:
```
$ docker inspect docker.redpanda.com/redpandadata/redpanda:v24.2.11 |jq -r '.[].Config.ExposedPorts'
{
  "8081/tcp": {},
  "8082/tcp": {},
  "9092/tcp": {},
  "9644/tcp": {}
}
```